### PR TITLE
[MRESOLVER-7] Download dependency POMs in parallel v2

### PR DESCRIPTION
**Pull request by slachiewicz**
Rebased version of@hwellmann pull #2 

I do not know if it's faster now, but I managed to compile the Maven project successfully and the it-core-tests completed correctly.

### CODE REVIEW

**Comment by slachiewicz** [comment]
@michael-o what changes i should move? all 3 commits?

**Comment by michael-o** [comment]
The subpackage and version bumps.

**Comment by slachiewicz** [comment]
ok updated #10 and new #11 #12

**Comment by mikehearn** [comment]
It looks like this has bit-rotted again. It'd be really nice to get this in.

Some classes have moved into a sub-package in #12, so now package-private visibilities are getting in the way. I wanted to try it out and see how much difference to speed it made, so I just quickly undid #12. 

For me it seems to roughly double the speed of artifact fetch. It'll probably make a much bigger difference on higher latency links. Nice work!

**Comment by michael-o** [comment]
@mikehearn What needs to be changed to make classes accesible for this PR? I'll change it.

**Comment by slachiewicz** [comment]
strange, but You requested to move #11 to diffrent PR. So now You need to revert #11 then you can apply again this PR..

**Comment by michael-o** [comment]
@slachiewicz I maybe wrote gibberish, I am working on five projects in parallel and simply lost track. Don't take my request as imperative.

**Comment by slachiewicz** [comment]
and I’m just trying to help:)

**Comment by steve-todorov** [comment]
Hi guys, are there any plans for this to be merged anytime soon? We have a bit project which might benefit from this. Is there anything more needed to be done?

**Comment by michael-o** [comment]
@steve-todorov Did you try the patch yourself? Unfortunately, there is a very little amount of people working on Resolver right now. I am personally too busy. There is so much new code that I have trouble to understand it, especially because I am not fully comfortable with the Resolver code.

@slachiewicz Can you rebase again?

**Comment by slachiewicz** [comment]
I did rebase code and I compile few projects with clean repo - it looks good and fast.
@steve-todorov could you try to test with Your projects also?

**Comment by mikehearn** [comment]
For what it's worth I've been testing it locally a fair bit and it's been working fine. However I had to fork Maven Resolver and add a few commits on top to get it working as well as I wanted it to:

https://github.com/mikehearn/maven-resolver/commits/master

Specifically, I had to roll back a change that conflicted with this PR, and then I added this commit:

https://github.com/mikehearn/maven-resolver/commit/1e2ed26e5c6b7f439882339deae8ff4effec50ac

to avoid creating non-daemon threads.

**Comment by slachiewicz** [comment]
Thx @mikehearn for tests and suggestion to use cachedThreadPool. I made small modification to use 5 core threads and max to 10 thread. Could you test new version?

**Comment by michael-o** [comment]
Can someone tell why the daemon threads along with the cached pool is necessary?

**Comment by slachiewicz** [comment]
ok, finaly i reuse in f1954149b7f4eb7ec7ccad6591c78702c959975e existing code used to create threads (org.eclipse.aether.util.concurrency.WorkerThreadFactory) in org.eclipse.aether.internal.impl.DefaultMetadataResolver#getExecutor

**Comment by michael-o** [comment]
Is this one now complete and can be reviewed again by me?

**Comment by slachiewicz** [comment]
yes, it’s final version :)

**Comment by michael-o** [comment]
I will drop cf1a1f7 because this is a different issue.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java [hunk 1]
```diff
@@ -91,6 +98,13 @@
 
     private VersionRangeResolver versionRangeResolver;
 
+    private static final ThreadGroup THREAD_GROUP = new ThreadGroup( "Maven Resolver dependency resolution" );
+
+    /**
+     * Setup executor with 5 daemon threads in pool, max to 10 concurrent and keep alive to 10 sec inactivity.
+     */
+    private ExecutorService executor = createExecutorService( 5, 10, 10 );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: Last arg is `10L`.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java [hunk 2]
```diff
@@ -91,6 +98,13 @@
 
     private VersionRangeResolver versionRangeResolver;
 
+    private static final ThreadGroup THREAD_GROUP = new ThreadGroup( "Maven Resolver dependency resolution" );
+
+    /**
+     * Setup executor with 5 daemon threads in pool, max to 10 concurrent and keep alive to 10 sec inactivity.
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: `10 sec` to > `10 s`.

**Comment by michael-o** [comment]
Maven master fails with:

    [INFO] Running org.apache.maven.repository.MirrorProcessorTest
    [INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in org.apache.maven.repository.MirrorProcessorTest
    [INFO]
    [INFO] Results:
    [INFO]
    [ERROR] Failures:
    [ERROR]   ProjectInheritanceTest.testDependencyManagement:68 Set size should be 4, is 3
    [INFO]
    [ERROR] Tests run: 142, Failures: 1, Errors: 0, Skipped: 0

Please review.

**Comment by michael-o** [comment]
@slachiewicz I'd like to publish MRESOLVER in a couple of days. Can you have a look at the PR again please?

**Comment by slachiewicz** [comment]
That use of DDCC is hard to remove without extending resolver-api and its jhard to understand how this work. I’m able to reproduce this bug but no idea about reason. I’ll look into this more later.

**Comment by slachiewicz** [comment]
Testcase: 
org.apache.maven.project.inheritance.t06.ProjectInheritanceTest#testDependencyManagement
and poms in maven-compat/src/test/resources/inheritance-repo/t06

Expected:
Artifact: maven-test:t06-a:jar 1.0 scope=compile optional=false
Artifact: maven-test:t06-b:jar 1.0 scope=compile optional=false
Artifact: maven-test:t06-d:jar 1.0 scope=test optional=false
Artifact: maven-test:t06-c:jar 1.0 scope=compile optional=false

Actual:
Artifact: maven-test:t06-a:jar 1.0 scope=compile optional=false
Artifact: maven-test:t06-b:jar 1.0 scope=compile optional=false
Artifact: maven-test:t06-c:jar 1.0 scope=compile optional=false

**Comment by michael-o** [comment]
While I understand MRESOLVER-9 I do not understand this failure. How can this happen if we don't have any changes in dep resolution, do we? Do I need to merge MRESOLVER-9 first?

**Comment by michael-o** [comment]
@slachiewicz, do you want to take another look at it. We want to roll 1.3.1 next week. I'd be cool if we could include this. My last questions are still open.

**Comment by michael-o** [comment]
@slachiewicz, do you think you can take on this? You know the change best by now.

**Comment by slachiewicz** [comment]
Unfortunately not for this release - I need more time to isolate a problem.

**Comment by gdubicki** [comment]
Hi @slachiewicz ! Were you able to get back to this? A lot of people would love to get this PR merged, see https://stackoverflow.com/questions/32299902/parallel-downloads-of-maven-artifacts/43832592 . :)

**Comment by JulienDemaziere** [comment]
Hi @slachiewicz, I'm very interested in this PR, what is blocking it ? What can I do to help ?

**Comment by michael-o** [comment]
@juliendemaziere. Time issues. You can work on it if you like, I'll be happy to review it.

**Comment by hwellmann** [comment]
FYI, I've looked at this issue again after all this time, and I've got it working now locally, including the entire maven-integration-testing suite. 

It looks like the missing bit is indeed related to https://issues.apache.org/jira/browse/MRESOLVER-9. I'm going to submit a new PR based on this one.

Thanks to @slachiewicz and everyone else who kept the candle burning!

**Comment by michael-o** [comment]
@hwellmann, what exact missing bit?

**Comment by hwellmann** [comment]
The missing bit: see this commit https://github.com/apache/maven-resolver/pull/30/commits/884928f82ac43dad570af15f7896d0a5cf54d2d6 in the new PR #30.

**Comment by michael-o** [comment]
@hwellmann, why is #30 necessary for this PR?

**Comment by hwellmann** [comment]
It fixes the failing test org.apache.maven.project.inheritance.t06.ProjectInheritanceTest observed by @slachiewicz in October 2018 as well as some failing tests from maven-integration-testing when using a build of Maven with this verion of Maven Resolver.

**Comment by michael-o** [comment]
Can this one be closed in favor of #30?

**Comment by slachiewicz** [comment]
Yes, of course.

**Comment by jira-importer** [comment]
Resolve #1397

**Comment by jira-importer** [comment]
Resolve #817

**Pull request closed without merging**